how-to/customize-workspace/client/src/connections.ts

```diff
@@ -0,0 +1,229 @@
+			return true;
+		}
+	}
+	return false;
+}
+
+async function executeAction(identity: OpenFin.Identity, payload: { action: string }): Promise<boolean> {
+	const actionId = payload?.action;
+	if (isActionSupported(identity, payload)) {
+		const requestedAction = registerAction[actionId];
+		console.log(`Executing action: ${actionId} on behalf of connection: ${identity.uuid}`);
+		await requestedAction();
+		return true;
+	}
+
+	console.log(
+		`Not Executing action: ${actionId} on behalf of connection: ${identity.uuid} as the action was either not provided or not listed in the supported actions definition.`
+	);
+	return false;
+}
+
+async function disconnect(identity) {
+	delete connectedClients[identity.uuid];
+	const connectedClientIds = Object.keys(connectedClients);
+	// disconnect from channel?
+	if (connectedClientIds.length === 0) {
+		// this logic can be updated to not disconnect or disconnect after a window of time
+		console.log(`No connections left.`);
+	}
+}
+
+export async function init(options: ConnectionProviderOptions) {
+	if (initialized) {
+		return;
+	}
+	initialized = true;
+	connectionOptions = options;
+	if (connectionOptions?.connectionId !== undefined) {
+		connectionService = await fin.InterApplicationBus.Channel.create(connectionOptions.connectionId);
+		console.log(`Configuring connection provider.`);
+		connectionService.onConnection((identity, payload) => {
+			// can reject a connection here by throwing an error
+			console.log("Client connection request identity:", JSON.stringify(identity));
+			console.log("Client connection request payload:", JSON.stringify(payload));
+
+			const validatedConnection = connectionOptions.connections.find(
+				(entry) => entry.identity.uuid === identity.uuid
+			);
+			let isValid = false;
+			let errorMessage =
+				"This connection has failed the validation check and cannot connect to the requested application.";
+			if (validatedConnection !== undefined) {
+				isValid = true;
+				if (validatedConnection.validatePayload) {
+					console.warn(
+						`This connection has specified payload validation but that check needs to be implemented. UUID: ${validatedConnection.identity.uuid}`
+					);
+					isValid = true;
+				}
+				if (isValid) {
+					if (connectedClients[identity.uuid] !== undefined) {
+						isValid = false;
+						errorMessage =
+							"This platform can only accept one connection from an external connection. The uuid is already registered.";
+					} else {
+						// assign the passed identity
+						validatedConnection.identity = identity;
+						connectedClients[identity.uuid] = validatedConnection;
+						console.log(`The following connection has been added to the connected list: ${identity.uuid}`);
+					}
+				}
+			}
+
+			if (!isValid) {
+				console.warn(
+					`The following connection has not been added to the connected list: ${identity.uuid} as it failed validation`
+				);
+				throw new Error(errorMessage);
```

Comment:
Should this really throw, who is going to catch it ?

Reply:
This is how you tell the connecting app that it can't connect. You have to throw